client/src/pages/Historic.jsx

```diff
@@ -1,11 +1,216 @@
-import React from 'react'
+import React, { useState, useEffect } from "react";
+import { useTheme } from '@mui/material/styles';
+import Box from '@mui/material/Box';
+import FirstPageIcon from '@mui/icons-material/FirstPage';
+import IconButton from '@mui/material/IconButton';
+import KeyboardArrowLeft from '@mui/icons-material/KeyboardArrowLeft';
+import KeyboardArrowRight from '@mui/icons-material/KeyboardArrowRight';
+import LastPageIcon from '@mui/icons-material/LastPage';
+import Paper from '@mui/material/Paper';
+import Table from '@mui/material/Table';
+import TableBody from '@mui/material/TableBody';
+import TableCell from '@mui/material/TableCell';
+import TableContainer from '@mui/material/TableContainer';
+import TableFooter from '@mui/material/TableFooter';
+import TableHead from '@mui/material/TableHead';
+import TablePagination from '@mui/material/TablePagination';
+import TableRow from '@mui/material/TableRow';
+import Typography from '@mui/material/Typography';
+
+const Historic = (props) => {
+  const { data, msToDate } = props
+  const { web3 } = data
+  const [allDonations, setAllDonations] = useState(null)
+  const [allProjects, setAllProjects] = useState(null)
+  const [allDonors, setAllDonors] = useState(null)
+
+  const [page, setPage] = useState(0);
+  const [rowsPerPage, setRowsPerPage] = useState(10);
+
+  useEffect(() => {
+    getAllDonations()
+  }, []);
+
+  useEffect(() => {
+    if (allDonations) {
+      console.log('allDonations :>> ', allDonations);
+    }
+  }, [allDonations]);
+
+  useEffect(() => {
+    if (allProjects) {
+      console.log('allProjects :>> ', allProjects);
+    }
+  }, [allProjects]);
+
+  useEffect(() => {
+    if (allDonors) {
+      console.log('allDonors :>> ', allDonors);
+    }
+  }, [allDonors]);
+
+  const findProjectInfos = (id, info) => {
+    for (const key in allProjects) {
```

Comment:
Pas nécessaire mais à savoir que dans les fonctions findXxxInfos on pourrait remplacer les boucles for par un array.find
https://www.w3schools.com/jsref/jsref_find.asp

Reply:
C'était ma 1ère idée, avant de passer par une fonction isolée je voulais tout regrouper dans getAllDonations() en mixant async et find() mais c'était plus simple de gérer de l'async dans un forin.  Au final j'ai isolé la fonction et retourné un tableau complet sans async.